src/cmd/linuxkit/moby/build.go

```diff
@@ -427,6 +453,53 @@ func tarAppend(iw *tar.Writer, tr *tar.Reader) error {
 	return nil
 }
 
+// Attempt to decompress a Linux kernel image
+// arm64 kernel images are plain gzip
+// x86 kernel images are in bzImage format, where the offset for the gzip'ed image is described in the header
+// s390x kernels images are not compressed.
+func decompressKernel(src *bytes.Buffer) (*bytes.Buffer, error) {
+	const gzipMagic = "\037\213\010" // Note, this is for gzip with max decoding used by the kernel
+
+	dst := new(bytes.Buffer)
+	s := src.Bytes()
+
+	if bytes.HasPrefix(s, []byte(gzipMagic)) {
+		log.Debugf("Found gzip signature at offset: 0")
+	} else {
+		// Not a plain gzip image, assume a bzImage and try to locate the compressed image
+		const setupSectorsIdx = 0x1f1
+		const sectorSize = 512
+		const payloadIdx = 0x248
+		const payloadLengthIdx = 0x24c
```

Comment:
Perhaps link to https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/x86/boot.txt for these?

Reply:
Also, I reckon you should test 0x1fe and 0x202:
```
01FE/2	ALL	boot_flag	0xAA55 magic number
0200/2	2.00+	jump		Jump instruction
0202/4	2.00+	header		Magic signature "HdrS"
```
for the `AA55` and `HdrS` magic numbers and if not find conclude this isn't an x86 bzImage at all.

Otherwise the behaviour of this code on uncompressed or non-x86 kernels is a little surprising I think.

You might optionally want to test the version header >= 2.08 which indicates the presence of the fields you consult.

AIUI we do want to support cross building images (or at least not make it harder to do in the future), so we can't use `runtime.GOARCH` to make decisions, right?

Reply:
Yes, there is demand for cross build at some point.

Reply:
> Otherwise the behaviour of this code on uncompressed or non-x86 kernels is a little surprising I think.

I don't understand this comment. If the user asks to decompress the kernel and we can't decompress it (or if it is not compressed), I think it's fair to error out. This should work for the arm64 LinuxKit kernels and x86 bzImage kernels. For s390x I don't think they are compressed, so don't specify `--decompress-kernel`...

> AIUI we do want to support cross building images (or at least not make it harder to do in the future), so we can't use runtime.GOARCH to make decisions, right?

I explicitly avoid this so the code should work on support kernel formats irrespective of the architecture we run `linuxkit build` on.

Reply:
> I don't understand this comment. If the user asks to decompress the kernel and we can't decompress it (or if it is not compressed), I think it's fair to error out

Right, but the error would have been something about no gzip header in the middle of a file, found via reading some offset which wasn't actually an offset just random bytes. This looks better now though.

I think the only remaining issue is that won't it panic on an array/slice overrun if the file is too short to have the header bytes in it, or if the file is corrupt such that payload offset points beyond the limits of the file?

Reply:
I'm fine with panicking if the kernel image is not large enough or corrupted, but added some check now.